### ISSUE TITLE
Discussion Thread Milestone B(2) - Display message from Field Technician

### ISSUE THREAD

**Issue by bsaiken** [FWAJL/FieldWorkManager#804]
Incoming messages, send from the Field Technician, is part of another Job.  The incoming message shall be stored in the message database table.  For testing purposes, we will manually add a Technician message to the database.  

The incoming messages shall be stored in the message history (discussion_content table) and the values of the following columns will be:
discussion_content_category_type (e.g. pm_id) 
discussion_content_category_value (id value of the discussion_content_category_type ) 

The stored message shall appear in the Message window.

**Comment by bsaiken** [comment]
@nadirhamid 
You cannot add the `status: ready to start` label.
I decide if Issue are ready to start.

**Comment by damirius** [comment]
@bsaiken @FWAJL 

I mentioned in the previous issue that I will implement history chat box here with the messages from services/FTs since they are the same (I'll later implement file attachments for the service users).
Just want to see if we want messages from the author (PM) there also?

**Comment by bsaiken** [comment]
@damirius @FWAJL 

Damir, I just wanted to make a general comment about this SOW.  I wrote a very specific and defined SOW because I had thought I was hiring someone new to the project.  Since you are on the job, and working hourly, I am more flexible in how this tool gets implemented.

I am confused about your statement that services/FTs are the same.  What does that mean?

Here is another thing to think about.  It would be beneficial to have the PM be able to carry on a communication with anyone associated with the Task.  By "anyone", I mean if they have an email address and/or a cell phone.  I have identified two parties: the (Task) Field Technicians; the (Task) Service Providers.  Eventually I would also like to include the (Project) Client.

Since on of the main purposes of this tool is to have an archive of all communications associated with a Task, we will want outgoing and incoming messages displayed and archived.  This will happen during the execution of the Task.  

Another thing to keep in mind is that once the project is "Complete", all the communications will get archived with the completed project.  The difference is, there can be no more messages added and no edits.  We just display static text.

Now, perhaps this answers your question directly, I have suggested a way to display the new message content and the history, but you are welcome to propose anything that makes sense.  

I like the message thread utilities on both eLance and github.  I think they inspired me to add this to my site.  Both, though, are more complex than this needs to be.  At least for this initial stage of development.  So, just let me see what you come up.

**Comment by damirius** [comment]
@bsaiken @FWAJL 

When I say that FT/Services are the same, I'm talking about our current db design regarding discussions. Which is a good thing. Discussion is between two users. If you recall when we redesigned our permission model we changed db tables, so services,FTs,PMs,clients are just users with some additional information stored in their respective tables. 
So currently you can have discussion between any two task users, which gives us huge flexibility in the coding. For example it will be very easy to add new user type in the task list, like client and PM could start discussion with him too.
I'll of course add some additional functionality to service users, like sending file links.

**Comment by damirius** [comment]
@bsaiken @FWAJL 

I've added chatbox on communication tab of active task view.
Since we changed the discussion table structure, I'll try to explain here how to add new "incoming" message.
Go to task list and choose some active task to view. Go on communications tab.
Check your `discussion` table in the DB. If there are some existing discussions they will be stored there.
Go back in your browser, choose service or FT from the list and click on the promote arrow to set it as the current discussion.
Go back to DB, you should see new discussion entry if this is the first time you started discussion between you (pm) and chosen FT/Service user. Remember `discussion_id` for that entry!
Go to `discussion_person` table, you should find two new entries for the new discussion, you can check them by `discussion_id`, you'll see that they have `user_id` fields. One of those two is the user_id of the PM, which is the author, you can see by checking `discussion_person_is_author` field. If it is set to 1, then that person is the PM for that Task.
So take a look at other of the two users (by checking discussion_id), write down `discussion_person_id`.
Go to `discussion_content` table, copy that `discussion_person_id` into appropriate field in that table and write the message under `discussion_content_message` field. ID and time fields will be automatically populated. You can add multiple entries if you wish.
Go back to your browser, you should be on the communications tab of active task with the current discussion selected. Refresh the page, you will see chatbox fill up with the messages from the FT/Service user (depending which one you selected). You can try sending them response in the upper text area box. Type whatever you wish and click send, note that maxchar limit in the db is 500.
You should see your message popping up in the chatbox without page refresh.

**Comment by bsaiken** [comment]
@damirius @

Perfect.  Just what I wanted.  I am moving this to code review, but please review the comments below.

Jeremie, once the code review is finished, please notify me when the live database is updated to accommodate this function.  I used Damir's reset script for testing, but then all my data entries were lost, including some users.  So, I will revert back to the old database for now.
  
There is one thing that needs to be fixed. I would like both your input on the best way to handle it.  As it is, it is possible to have technicians, service providers, or clients that are not users (that is to say, they are not in the `user` table).  So, they could show up on the communications tab list as potential message recipients, but it does not work.  There are several reasons why someone would not be a user.  For example, If I add a new user, say a technician, and I use a duplicate email address, we are currently not screening for that and they are added as a technician, but not a new user.  We are not requiring an email address to add a new FT, client, or Service Provider, so these could be added, but would not be users.  

Eventually, once a person is added as a user, we will want to first send them a confirmation email that they must reply to before they are added.  If we tried to start a conversation before they confirmed, they would not be a user yet.

I propose this fix: we add a popup confirm box alerting the PM that their selected recipient is not a registered user.  The confirm box would have three buttons: `Cancel`, `Waiting for user confirmation` or `Add as User`.  If they select `Cancel`, there is no message.  `Waiting for user confirmation` stores the message(s) until the user confirms, then all stored messages are sent at once.  `Add as user` runs the script to add them as a user.  This requires that the email field be populated, and if it is not, we have a prompt box asking for their email address (and we check that it is unique).

Thoughts?

**Comment by FWAJL** [comment]
@damirius 

It doesn't work for me. I have added a new FT in order to have a user associated with him. Still, it says "Discussion couldn't be start". See below:
![screen shot 2015-04-27 at 06 33 43](https://cloud.githubusercontent.com/assets/7524388/7341782/d57f7b7c-eca8-11e4-8e4e-24ec8139ad4a.png)

My first thoughts about your comments @bsaiken is that we can't avoid not creating the user row for each technician, project_manager, client, service created. The more we keep going, the worst it is going to be and we will spend more time making tweaks like the one described than actually what is missing. 

I think we must 

1. have the email value in the table user instead of all the other tables and we set a unique constraint there to avoid duplicates. Then the email will be truly unique whereas it wouldn't if stored in other tables.
2. modify the code where the former email column was used
3. make the email required everywhere.
4. implement the user creation for each technician, project_manager, client, service created

That should be done before continuing this discussion thread work.

Let me know your thoughts. Thank you.

**Comment by bsaiken** [comment]
@fwajl
I had to run Damir's database reset for it to work. 

We need to move forward with the discussion thread development and not delay.  Dealing with users can wait. We can add this to our discussion Wednesday.

**Comment by bsaiken** [comment]
@FWAJL 
I had to run Damir's database reset for it to work.

Please see my comment, above.

**Comment by damirius** [comment]
@FWAJL @bsaiken 

I see that our priority is still discussion milestone here, but we are still running into a lot of obstacles. For example this problem with users. When I was working on our role system, we were talking about all types having user entry in the db. So this stands for FTs, Services, Clients, PMs. 
Discussion DB table design was built on that assumption and I think it will serve our purpose better than the old design.
Now this isn't the only thing of course.
We still need Service user pages if we want chat box for them too, which I assume we do. Since I see that we need a way to send messages with attachments as the part of the issue #806.
We also need that popup with list of available forms so we can use it as the file attachment option for both services and pms.
Also I remembered that we also need email class wrapper which we will use for sending notifications and/or whole messages directly to the emails of FTs,Services and PMs, which is one of the reason for this addition of valid emails to the user table.

**Comment by bsaiken** [comment]
@FWAJL @damirius 
I think my comment for #805 will answer this.  We only communicate with "registered" users - e.g. - those with a user entry in the database.  Then, we deal with getting the user into the database somewhere else.

If we need to split this into separate Issues then let's do it.

**Comment by WebDevJL** [comment]
@damirius @bsaiken 

Issue #916 is taking care of adding a user row for PM, FT, Client and Service provider. What Damir for the technician is reused. I just made it generic. The only thing not done is making the Service provider email mandatory.

Also, Damir, do you understand how we want to code the attachments part of the Discussion work?

**Comment by damirius** [comment]
@FWAJL 

If I'm correct we will have 2 types of attachments. One is regular upload of files from the local machine, and the other is choosing already existing files on the server which, if I'm not mistaken, are those automatically copied forms for that task.

**Comment by WebDevJL** [comment]
@bsaiken @damirius 

Yes, those are the sources of attachments. What I would like to know also is if you understand how attachments will be included in the message/email sent.

Basically, once the form(s) are selected or document(s) are uploaded, we will create "download" links to those files so the recipient(s) can get the files in a browser. So we are not using the attachment as email have but rather a more flexible download link attachments method.

What do you think?

**Comment by damirius** [comment]
@FWAJL @bsaiken 

Yea, that's easiest, since we still don't know if we will use email for sending actual messages, plus showing download link will be great for chatbox.

**Comment by bsaiken** [comment]
@damirius @WebDevJL 
With fancybox, we can view the pdf files, even do some entry (I think).  But, downloading is important, especially for the Field Technicians. If they are at a remote site, they can just fill out downloaded files and upload them later, when they have internet access again.

**Comment by bsaiken** [comment]
@damirius @WebDevJL 
I don't necessarily want to send messages via email, but it's sort of moot since, without email, the user wouldn't know there was a message unless they logged in and looked.  And I can tell you, the way field work goes, that is not practical.  Emailing will be necessary.

**Comment by bsaiken** [comment]
@damirius 
Is this ready for code review?

**Comment by damirius** [comment]
@bsaiken @FWAJL 

I believe that we already done that but we run into that problem with non existing users for services/fts because of invalid emails and what not. So we set it on hold till we fix that.

**Comment by bsaiken** [comment]
@damirius @FWAJL 
OK - I am clearing out the database.  I'll post here when I am done.

**Comment by bsaiken** [comment]
@damirius 
The database is purged.  Only one FT per PM now.  I may have deleted the login you had been using.  Use `mapping`.  

Would this work if I inserted the necessary entries into the database?

**Comment by damirius** [comment]
@bsaiken

It should

**Comment by bsaiken** [comment]
@damirius @FWAJL 
I successfully tested this by inserting the reply into the database (since the Mobile UI is not yet working).
Note that because of the manual insertion, I had to log out and log back in to see the reply.
![screen shot 2015-05-06 at 8 00 17 am](https://cloud.githubusercontent.com/assets/8228624/7495955/1e84ef90-f3c6-11e4-9acc-2f6b99f33013.png)

**Comment by FWAJL** [comment]
@bsaiken 

This is requiring issue #850. On hold for now.
